README.md
Abigail Log Plugin
---

<p align="right">
  <a href="https://npmjs.org/package/abigail-plugin-log">
    <img src="https://img.shields.io/npm/v/abigail-plugin-log.svg?style=flat-square">
  </a>
  <a href="https://travis-ci.org/abigailjs/abigail-plugin-log">
    <img src="http://img.shields.io/travis/abigailjs/abigail-plugin-log.svg?style=flat-square">
  </a>
  <a href="https://codeclimate.com/github/abigailjs/abigail-plugin-log/coverage">
    <img src="https://img.shields.io/codeclimate/github/abigailjs/abigail-plugin-log.svg?style=flat-square">
  </a>
  <a href="https://codeclimate.com/github/abigailjs/abigail-plugin-log">
    <img src="https://img.shields.io/codeclimate/coverage/github/abigailjs/abigail-plugin-log.svg?style=flat-square">
  </a>
  <a href="https://gemnasium.com/abigailjs/abigail-plugin-log">
    <img src="https://img.shields.io/gemnasium/abigailjs/abigail-plugin-log.svg?style=flat-square">
  </a>
</p>

No installation
---

> abigail built-in plugin

Usage
---
if this plugin is enabled, output the debugging information.

```bash
abby test
# +    1 ms @_@ use package.json.
# +    1 ms @_@ plugin enabled exit, log, launch, watch.
# +   21 ms @_@ task start test.
# ...
#   18 passing (56ms)
# ...
# +  2.8  s @_@ task end test. exit code 0.
# +    1 ms @_@ ... watch at src/**/*.js, test/**/*.js.
```

if turned off, becomes only to the output of the emulator.

```bash
abby test --no-log
# ...
#   18 passing (56ms)
# ...
```

See also
---
* [abigailjs/abigail](https://github.com/abigailjs/abigail#usage)
* [abigailjs/abigail-plugin](https://github.com/abigailjs/abigail-plugin#usage)

Development
---
Requirement global
* NodeJS v5.7.0
* Npm v3.7.1

```bash
git clone https://github.com/abigailjs/abigail-plugin-log
cd abigail-plugin-log
npm install

npm test
```

License
---
[MIT](http://abigailjs.mit-license.org/)


CHANGELOG.md
<a name="0.3.0"></a>
# [0.3.0](https://github.com/abigailjs/abigail-plugin-log/compare/v0.2.1...v0.3.0) (2016-04-09)


### Features

* **scriptSuffixes:** add behavior `output scriptSuffixes` ([079c283](https://github.com/abigailjs/abigail-plugin-log/commit/079c283))



<a name="0.2.1"></a>
## [0.2.1](https://github.com/abigailjs/abigail-plugin-log/compare/v0.2.0...v0.2.1) (2016-04-09)


### Bug Fixes

* remove uglify, plugin name was broken ([22510b6](https://github.com/abigailjs/abigail-plugin-log/commit/22510b6))



<a name="0.2.0"></a>
# [0.2.0](https://github.com/abigailjs/abigail-plugin-log/compare/v0.1.1...v0.2.0) (2016-04-09)


### Features

* follow the `initialized` lifecycle ([29cd1c1](https://github.com/abigailjs/abigail-plugin-log/commit/29cd1c1))



<a name="0.1.1"></a>
## [0.1.1](https://github.com/abigailjs/abigail-plugin-log/compare/06ecaf2...v0.1.1) (2016-03-30)


### Bug Fixes

* missing script name property ([e30ccf8](https://github.com/abigailjs/abigail-plugin-log/commit/e30ccf8))
* typo ([abc9797](https://github.com/abigailjs/abigail-plugin-log/commit/abc9797))
* **task-start:** fix abigailjs/abigail#9 ([1f0d57a](https://github.com/abigailjs/abigail-plugin-log/commit/1f0d57a)), closes [abigailjs/abigail#9](https://github.com/abigailjs/abigail/issues/9)

### Features

* add spec `fatal event: should be output with the iconFatal` ([06ecaf2](https://github.com/abigailjs/abigail-plugin-log/commit/06ecaf2))





test/index.js
// dependencies
import AsyncEmitter from 'carrack';
import { lookupSync } from 'climb-lookup';
import sinon from 'sinon';
import stripAnsi from 'strip-ansi';
import assert from 'power-assert';

// target
import Log from '../src';

// fixture
const createMockProcess = () => ({
  cwd: () => process.cwd(),
  exit: sinon.spy(),
  stdout: {
    write: sinon.spy(),
  },
});
const setupFixtures = (options = {}) => {
  const emitter = new AsyncEmitter;
  const process = createMockProcess();
  const opts = { process, notifyCwd: false, notifyPlugins: false, ...options };
  const log = new Log(emitter, true, opts);

  log.setProps({
    json: {
      path: lookupSync('package.json'),
    },
  });

  return { log, emitter, process };
};

// specs
describe('Log', () => {
  describe('plugin lifecycle', () => {
    describe('attach-plugins', () => {
      it('should be output the relative packagePath', () => {
        const { emitter, process } = setupFixtures({ notifyCwd: true });

        return emitter.emit('initialized')
        .then(() => emitter.emit('attach-plugins'))
        .then(() => {
          const message = stripAnsi(process.stdout.write.args[0][0]);

          assert(process.stdout.write.callCount === 1);
          assert(message.match(/use package.json\.\n$/));
        });
      });

      it('should be output that package.json is missing', () => {
        const { log, emitter, process } = setupFixtures({ notifyCwd: true });
        log.setProps({ json: null });

        return emitter.emit('initialized')
        .then(() => emitter.emit('attach-plugins'))
        .then(() => {
          const message = stripAnsi(process.stdout.write.args[0][0]);

          assert(process.stdout.write.callCount === 1);
          assert(message.match(/missing package.json\.\n$/));
        });
      });

      it('should be output the available plugin names', () => {
        const { log, emitter, process } = setupFixtures({ notifyPlugins: true });
        log.setProps({
          plugins: {
            foo: { name: 'foo' },
            bar: { name: 'bar' },
            baz: { name: 'baz' },
          },
        });

        return emitter.emit('initialized')
        .then(() => emitter.emit('attach-plugins'))
        .then(() => {
          const message = stripAnsi(process.stdout.write.args[0][0]);

          assert(process.stdout.write.callCount === 1);
          assert(message.match(/plugin enabled foo, bar, baz\.\n$/));
        });
      });

      it('should be output the parent.props.scriptSuffixes(abigail#12)', () => {
        const { log, emitter, process } = setupFixtures({ notifyPlugins: true });

        const tasks = [
          {
            main: {
              name: 'foo',
              exitCode: 1,
            },
          },
          {
            main: {
              name: 'bar',
              exitCode: 2,
            },
          },
          {
            main: {
              name: 'baz',
              exitCode: 3,
            },
          },
        ];
        log.setProps({
          scriptSuffixes: ['beep', 'boop', 'kaboom'],
        });

        return emitter.emit('initialized')
        .then(() => emitter.emit('attach-plugins'))
        .then(() => emitter.emit('task-start', tasks))
        .then(() => {
          const output = stripAnsi(process.stdout.write.args[1][0]);
          assert(output.match(/\+[ \d]+ms @_@ task start foo, bar, baz. \(with beep boop kaboom\)\n$/));
        });
      });
    });

    describe('detach-plugins', () => {
      it('if the exit code is 0, it should cheer', () => {
        const { emitter, process } = setupFixtures();

        return emitter.emit('initialized')
        .then(() => emitter.emit('attach-plugins'))
        .then(() =>
          emitter.emit('task-end', [{ script: { exitCode: 0 } }])
        )
        .then(() => emitter.emit('detach-plugins'))
        .then(() => {
          const message = stripAnsi(process.stdout.write.args[1][0]);

          assert(process.stdout.write.callCount === 2);
          assert(message.match('cheers for good work.\n'));
        });
      });

      it('unless the exit code is 0, it should apologize', () => {
        const { emitter, process } = setupFixtures();
        return emitter.emit('detach-plugins').then(() => {
          const message = stripAnsi(process.stdout.write.args[0][0]);

          assert(process.stdout.write.callCount === 1);
          assert(message.match("i'm terribly sorry...\n"));
        });
      });
    });

    describe('log', () => {
      it('should be output with the icon', () => {
        const { emitter, process } = setupFixtures();
        return emitter.emit('initialized')
        .then(() => emitter.emit('attach-plugins'))
        .then(() =>
          emitter.emit('log', 'bar!').then(() => {
            const output = stripAnsi(process.stdout.write.args[0][0]);
            assert(output.match(`ms ${stripAnsi(Log.icon)} bar!`));
          })
        );
      });
    });

    describe('script-error', () => {
      it('should be output with the icon', () => {
        const { emitter, process } = setupFixtures();
        return emitter.emit('initialized')
        .then(() => emitter.emit('attach-plugins'))
        .then(() =>
          emitter.emit('script-error', new Error('baz~')).then(() => {
            const output = stripAnsi(process.stdout.write.args[0][0]);
            assert(output.match(`ms ${stripAnsi(Log.iconFatal)} Error: baz~`));
          })
        );
      });
    });

    describe('script-start, script-end', () => {
      const task = {
        script: { name: 'foo' },
        exitCode: 1,
      };

      it('should be output the object.name unless single task', () => {
        const { emitter, process } = setupFixtures();
        return emitter.emit('initialized')
        .then(() => emitter.emit('attach-plugins'))
        .then(() =>
          emitter.emit('task-start', [{}, {}])
        )
        .then(() =>
          emitter.emit('script-start', task).then(() => {
            const output = stripAnsi(process.stdout.write.args[1][0]);
            assert(output.match(/\+[ \d]+ms @_@ script start foo.\n$/));
          })
        );
      });

      it('should be output the object.name and exitCode unless single task', () => {
        const { emitter, process } = setupFixtures();
        return emitter.emit('initialized')
        .then(() => emitter.emit('attach-plugins'))
        .then(() =>
          emitter.emit('task-start', [{}, {}])
        )
        .then(() =>
          emitter.emit('script-end', task).then(() => {
            const output = stripAnsi(process.stdout.write.args[1][0]);
            assert(output.match(/\+[ \d]+ms @_@ script end foo. exit code 1.\n$/));
          })
        );
      });

      it('should be output the object.name and exitCode in serial execution', () => {
        const { emitter, process } = setupFixtures();
        return emitter.emit('initialized')
        .then(() => emitter.emit('attach-plugins'))
        .then(() =>
          emitter.emit('task-start', [[
            [{ main: { raw: 'echo foo' } }],
            [{ main: { raw: 'echo bar' } }],
          ]])
        )
        .then(() =>
          emitter.emit('script-end', task).then(() => {
            const output = stripAnsi(process.stdout.write.args[1][0]);
            assert(output.match(/\+[ \d]+ms @_@ script end foo. exit code 1.\n$/));
          })
        );
      });
    });

    describe('task-start, task-end', () => {
      it('should be output with the icon', () => {
        const { emitter, process } = setupFixtures();
        const tasks = [
          {
            main: {
              name: 'foo',
              exitCode: 1,
            },
          },
          {
            main: {
              name: 'bar',
              exitCode: 2,
            },
          },
          {
            main: {
              name: 'baz',
              exitCode: 3,
            },
          },
        ];

        return emitter.emit('initialized')
        .then(() => emitter.emit('attach-plugins'))
        .then(() =>
          emitter.emit('task-start', tasks).then(() => {
            const output = stripAnsi(process.stdout.write.args[0][0]);
            assert(output.match(/\+[ \d]+ms @_@ task start foo, bar, baz.\n$/));
          })
        );
      });

      it('should be output with the icon', () => {
        const { emitter, process } = setupFixtures();
        const tasks = [
          {
            script: { name: 'foo' },
            exitCode: 1,
          },
          {
            script: { name: 'bar' },
            exitCode: 2,
          },
          {
            script: { name: 'baz' },
            exitCode: 3,
          },
        ];

        return emitter.emit('initialized')
        .then(() => emitter.emit('attach-plugins'))
        .then(() =>
          emitter.emit('task-end', tasks).then(() => {
            const output = stripAnsi(process.stdout.write.args[0][0]);
            assert(output.match(/\+[ \d]+ms @_@ task end foo, bar, baz. exit code 1, 2, 3.\n$/));
          })
        );
      });
    });

    describe('watch', () => {
      it('should be output with the icon', () => {
        const { emitter, process } = setupFixtures();
        return emitter.emit('initialized')
        .then(() => emitter.emit('attach-plugins'))
        .then(() =>
          emitter.emit('watch', '/path/to/dir', 'changed').then(() => {
            const output = stripAnsi(process.stdout.write.args[0][0]);
            assert(output.match(/\+[ \d]+ms @_@ file \/path\/to\/dir changed.\n$/));
          })
        );
      });
    });
  });

  describe('.strong', () => {
    it('if first argument is array, should be join using second argument', () => {
      const output = stripAnsi(Log.strong(['foo', 'bar', 'baz'], '! '));
      assert(output === 'foo! bar! baz');
    });
  });

  describe('.statuses', () => {
    it('if second argument is 1, the first argument should be in the fail style', () => {
      assert(Log.statuses('kaboom', 1) === Log.fail('kaboom'));
      assert(Log.statuses('kaboom', 0) === Log.pass('kaboom'));
    });

    it('if specify first argument is array, should return the string', () => {
      const glue = '!!! ';
      const expectedCharacters = [
        Log.pass('a'),
        Log.fail('b'),
        Log.fail('c'),
      ].join(glue);

      assert.equal(
        Log.statuses(['a', 'b', 'c'], [0, 1, 2], glue),
        expectedCharacters,
      );
    });

    it('second argument should be omitted', () => {
      assert(Log.statuses(0) === Log.pass('0'));
      assert(Log.statuses(1) === Log.fail('1'));

      const glue = 'ಠ益ಠ';
      assert.deepEqual(
        Log.statuses([0, 1, 329108], null, glue),
        [
          Log.pass('0'),
          Log.fail('1'),
          Log.fail('329108'),
        ].join(glue),
      );
    });
  });

  describe('.format', () => {
    it('should return the elapsed time in the 7 characters', () => {
      let time;

      time = Log.format(1000);
      assert(time === '1000 ms');
      assert(time.length === 7);

      time = Log.format(1500);
      assert(time === ' 1.5  s');
      assert(time.length === 7);

      time = Log.format(1000 * 60);
      assert(time === '  60  s');
      assert(time.length === 7);

      time = Log.format(1234 * 20);
      assert(time === '24.7  s');
      assert(time.length === 7);

      time = Log.format(1000 * 60 * 60);
      assert(time === '  60min');
      assert(time.length === 7);

      time = Log.format(1234 * 60 * 20);
      assert(time === '24.7min');
      assert(time.length === 7);

      time = Log.format(1000 * 60 * 60 * 24);
      assert(time === '  24 hr');
      assert(time.length === 7);
    });
  });

  describe('::output', () => {
    it('should output the arguments with elapsed time and abby icon', () => {
      const { log, process } = setupFixtures();

      log.output('foo!');
      const output = stripAnsi(process.stdout.write.args[0][0]);

      assert(output.match(/\+[ \d]+ms @_@ foo!\n$/));
    });
  });

  describe('::getElapsed', () => {
    it('should return the elapsed time until called for formatted 7 characters', () => {
      const { log } = setupFixtures();
      const elapsed = log.getElapsed();

      assert(elapsed.length === 7);
      assert(elapsed.match(/[ \d]+ms$/));
    });
  });
});


src/index.js
// dependencies
import Plugin from 'abigail-plugin';
import chalk from 'chalk';
import { relative as relativePath } from 'path';
import flattenDeep from 'lodash.flattendeep';

// @class Log
export default class Log extends Plugin {
  /**
  * @static
  * @property defaultOptions
  */
  static defaultOptions = {
    notifyCwd: true,
    notifyPlugins: true,
  }

  /**
  * @static
  * @property icon
  */
  static icon = chalk.magenta('@_@');

  /**
  * @static
  * @property iconFatal
  */
  static iconFatal = chalk.magenta.inverse('@_@;');

  /**
  * @static
  * @property pass
  */
  static pass = chalk.cyan.underline;

  /**
  * @static
  * @property fail
  */
  static fail = chalk.yellow.underline;

  /**
  * @static
  * @method important
  * @param {array|string} arg - an output string
  * @param {string} glue - a string separator
  * @returns {string} output - underlined characters
  */
  static important(arg, glue = ', ') {
    if (arg instanceof Array) {
      return arg.map(word => chalk.inverse(word)).join(glue);
    }
    return chalk.inverse(arg);
  }

  /**
  * @static
  * @method strong
  * @param {array|string} arg - an output string
  * @param {string} glue - a string separator
  * @returns {string} output - underlined characters
  */
  static strong(arg, glue = ', ') {
    if (arg instanceof Array) {
      return arg.map(word => chalk.underline(word)).join(glue);
    }
    return chalk.underline(arg);
  }

  /**
  * @static
  * @method statuses
  * @param {array|any} arg - source chalacters
  * @param {array|number} ref - exit code
  * @param {string} glue - a string separator
  * @returns {string} output - ansi colored characters (cyan or yellow)
  */
  static statuses(arg, ref, glue = ', ') {
    const args = arg instanceof Array ? arg : [arg];
    let refs;
    if (ref instanceof Array) {
      refs = ref;
    } else if (ref !== undefined && ref !== null) {
      refs = [ref];
    } else {
      refs = args;
    }

    return args.map((status, i) => {
      if (refs[i] > 0) {
        return Log.fail(status);
      }

      return Log.pass(status);
    }).join(glue);
  }

  /**
  * @static
  * @method
  * @param {number} diff - millseconds
  * @returns {string} elapsed - human readable time
  */
  static format(diff) {
    let weight = diff;
    let unit = ' ms';
    if (weight > 1000) {
      weight = Math.round((weight / 1000) * 10) / 10;
      unit = '  s';
      if (weight > 60) {
        weight = Math.round((weight / 60) * 10) / 10;
        unit = 'min';
        if (weight > 60) {
          weight = Math.round((weight / 60) * 10) / 10;
          unit = ' hr';
        }
      }
    }

    return `   ${weight}${unit}`.slice(-7);
  }

  /**
  * the plugin lifecycle method
  * execute only once before the parse
  *
  * @method pluginDidInitialize
  * @returns {undefined}
  */
  pluginDidInitialize() {
    this.subscribe('log', (...args) => {
      this.output(...args);
    });
    this.subscribe('script-error', (...args) => {
      this.outputFatal(...args);
    });

    const scriptStart = ({ script }) => {
      this.output(`script start ${Log.strong(script.name)}.`);
    };
    const scriptEnd = ({ script, exitCode }) => {
      const name = Log.statuses(script.name, exitCode);
      const code = Log.statuses(exitCode);

      this.output(`script end ${name}. exit code ${code}.`);
    };

    this.subscribe('task-start', (task = []) => {
      const names = flattenDeep(task).map((serial) => {
        if (serial.main) {
          return serial.main.name;
        }
        return 'unknown';
      });
      const suffixes = this.getProps().scriptSuffixes || [];
      let suffix = '';
      if (suffixes.length) {
        suffix = ` (with ${Log.strong(suffixes, ' ')})`;
      }
      this.output(`task start ${Log.strong(names)}.${suffix}`);

      if (names.length > 1) {
        this.subscribe('script-start', scriptStart);
        this.subscribe('script-end', scriptEnd);
      }
    });
    this.subscribe('task-end', (result = []) => {
      const scripts = flattenDeep(result);
      const names = scripts.map((script) => script.script.name);
      const codes = scripts.map((script) => script.exitCode);
      this.exit = scripts.reduce((prev, script) => {
        if (prev > 0 || script.exitCode > 0) {
          return 1;
        }
        return 0;
      }, 0);
      this.output(`task end ${Log.statuses(names, codes)}. exit code ${Log.statuses(codes)}.`);

      this.parent.removeListener('script-start', scriptStart);
      this.parent.removeListener('script-end', scriptEnd);
    });

    this.subscribe('watch', (path, event) => {
      this.output(`file ${chalk.bold(path)} ${event}.`);
    });
  }

  /**
  * @method pluginWillAttach
  * @returns {undefined}
  */
  pluginWillAttach() {
    this.elapsed = Date.now();

    if (this.opts.notifyCwd) {
      const json = this.getProps().json;
      if (json && json.path) {
        const path = relativePath(this.opts.process.cwd(), json.path);
        this.output(`use ${chalk.inverse(path)}.`);
      } else {
        this.output('missing package.json.');
      }
    }

    if (this.opts.notifyPlugins) {
      const plugins = this.getProps().plugins || {};
      const available = [];
      for (const key in plugins) {
        if (plugins.hasOwnProperty(key) === false) {
          continue;
        }
        available.push(plugins[key].name);
      }
      this.output(`plugin enabled ${Log.important(available)}.`);
    }
  }

  /**
  * @method pluginWillAttach
  * @returns {undefined}
  */
  pluginWillDetach() {
    if (this.exit === 0) {
      this.output('cheers for good work.');
    } else {
      this.outputFatal("i'm terribly sorry...");
    }
  }

  /**
  * @public
  * @method output
  * @param {any} args - output string
  * @returns {undefined}
  */
  output(...args) {
    const ms = this.getElapsed();
    const msColored = chalk.gray(`+ ${ms}`);
    this.opts.process.stdout.write(`${msColored} ${Log.icon} ${args.join(' ')}\n`);
  }

  /**
  * @public
  * @method outputFatal
  * @param {any} args - an output string
  * @returns {undefined}
  */
  outputFatal(...args) {
    const ms = this.getElapsed();
    const msColored = chalk.gray(`+ ${ms}`);
    this.opts.process.stdout.write(`${msColored} ${Log.iconFatal} ${args.join(' ')}\n`);
  }

  /**
  * @public
  * @method getElapsed
  * @returns {undefined}
  */
  getElapsed() {
    const diff = (Date.now() - this.elapsed) || 0;
    this.elapsed = Date.now();
    return Log.format(diff);
  }
}
